### ISSUE TITLE
Settings refactoring

### ISSUE THREAD

**Issue by Utumno** [wrye-bash/wrye-bash#178]
Grouping settings fixups and todos:

306:
- [x] Most code paths in bosh.InitSettings are broken - fixed in d4309cca4a2a08f1489bae2ec88ec84cf0b09780

307:
- saved load orders should NOT go to BashSettings !!! -> moved to #312
- [x] deprecate support for older versions, `.pkl` files etc ( #253 )
- [x] namespace settings ( cffffd2ae1c52a48610f6742dd1bed997f45d52a ):
  - [x] rename settings local variables
  - [x] drop bosh.settings -> bass.settings
  - [x] ~~drop basher.settings -> bass.settings~~
- [ ] hunt down dirs[] - necessary step to centralized settings location
- [x] ~~clean removed settings (for instance vScrollPos)~~ well, some other spring
- [ ] _always_ save settings - see 2914aac1c32f10441083ddba63080a33b839cd3c
- [ ]  cleanup defaults
- [ ] save settings of a tab on closing it (the tab only) - test Saves, Inis, Installers etc
- [ ] Bashed patch config handling refactor - as it is renaming the bashed patch looses its config. Save them separately and index them based on CRC
- [ ] from svn: "3) it does not do similar recovery actions for Table.dat/Table.dat.bak, meaning that even if BashSettings.dat is deleted, the program still would not start due to Table.dat being in the wrong format" - test this
- [x] reset "ask continue" menu item - must be done for all of them at once I guess - done as part of Settings dialog

308:

Settings are accessed from everywhere - we need to centralize this. Note for instance the TankData setPatams etc, the basher.settings global, the balt._settings etc etc etc. We need a package where also the backup classes will go. Especially in balt a class using `_settings` should be moved to a submodule - closer to basher (uses specific Bash stuff, while balt must have a clean wx wrapping part).

See also [Boot-sequence WIP page](https://github.com/wrye-bash/wrye-bash/wiki/%5BWIP%5D-Boot-sequence)

**Comment by Utumno** [comment]
@iaz3: replying to you here to not spam the loot pull req.

First of all I do not have time for profiles now - it's a full time feature and I am trying to release.

> Does bash read directly from the ini or cache it?
The ini is not supposed to be reloaded (that's one of the point ##: needs work refers to in bass.GetBashIni). The cache is supposed not to change at this moment

> But how does that work for different games? 

Settings handling by Bash is complex and there are many reasons why it is so. It tries to achieve as much modularity per game as possible which resulted in a bunch of settings in a bunch of places some of them configurable.

I will post in the profiles issue when I find some time but for now Profiles are blocked by this issue - that is settings refactoring. If you want to spend brain power spend it on **barb.py.**
To even start looking at profiles we need barb.py refactored to:

- [ ] have a global array of sorts that gives us (with the required path join operations) all possible Bash settings paths - had started working on that will post my no-op code
- [ ] drop the unpythonic design and replace it with a set of functions (Pep8 this time) that do exactly the same thing in a more modular manner. The functions should be simple and readable but their combination powerful (and tested) enough to manipulate _all_ Bash settings at once so we can copy them around to build profiles

Profiles implementation, scope and specification is still nebulous and needs a lot of thought (and user feedback). So start simple. We first need settings handling completely and reliably sorted. So if you want to start with profiles start with this issue - note how many commits refer to this to get an vague idea of what's been done - the OP here gives also some pointers but the issue is so complicated that i can't pin it down to a couple check boxes.

Actually settings (and therefore profiles) are the major goals for 308 - along with patcher testing. And wx3

----

Re: inno installer. Dunno if I can have this in for the beta but:

- I need help with current installer - if you can fix the issues reported by alt3rn1ty in the last posts of the forum thread it would be great help. It should be easy if you are familiar with nsi installers
- alternatively concentrate on the inno installer so we can replace the one we have - no profiles nothing just a drop in replacement for current one. Are you confident enough that you can pack this for the users to test ? Is package_for_release working perfectly (no room for errors here, that's packaging Bash we can't go wrong) ?

----

So bottom line - leave profiles please for now it's a very hard nut (or keep working but don't notify me) - keep your code and I'll have a look but not ASAP. If you start on them start from **barb.py**.

The inno installer would be nice to have - but you must make sure everything is fine - test, then pack, then post at the forums for testing. package_for release should work 100% correct.

If you can fix me the couple bugs of the nsi installer then please do this ASAP - I need to get the beta out and no installer == no beta.

----

Going back to rebasing my branch

**Comment by iaz3** [comment]
> The ini is not supposed to be reloaded (that's one of the point ##: needs work refers to in bass.GetBashIni). The cache is supposed not to change at this moment

So bad things will happen if the cache is changed

WAIT, as i'm typing this, i just realized. We could, in theory, simply change the ini based on the game selected BEFORE bash reads it? That would work, right? It's not optimal, but it should accomplish the same thing.(IE, the app structure would therefore be, `{app}\Bash Installers\{game name}` and `{app}\Bash Mod Data\{game name}` where {game name} is added in(ConfigParser?) at runtime)

Bash settings are stored relative to those directories, right? So changing where they point before the ini is read should effectively swap them out?

as for `package_for_release` any changes there were mostly for my convenience, it should work the same no matter what and i frequently use it to compare the outputs from the new Building\build.py(Which is where my new code is.)

I can probably rig up inno setup to support detecting and installing for different games. It has built in functions for getting registry values, so it shouldn't be too hard to define components based on installed games

I'll see what i can do on that, and have it install into {game}/mopy like normal, per game.

**Comment by Utumno** [comment]
Yep - in general I would like to have the inno set up to just replace the nsi one - so same logic, extra loactions and all. Don't forget that we should change docs and posts and turorials all over the place.

then when the dust settles we should adress the settings issue. When we have a powerful API to handle settings we can look at profiles. _But only then_. Also while refactoring settings you will get a much clearer picture of the code and how ini and all the other settings are used and all the complications so we can really discuss profiles

Profiles will be two steps:

- just game ones leaving all the rest in one place
- real profiles - much harder

Here is my profile handling for now:

```
REM fallout4.bat
del bash.ini
copy "_inis\bash - FALLOUT4.ini" bash.ini
start /B C:\_\python27\python.exe "Wrye Bash Launcher.pyw" -d &
REM pause
```

You get the picture ;)

**Comment by iaz3** [comment]
On the installer, the latest commit on that branch should be working pretty well.

From my basic testing, everything works correctly, it detects games and their paths just fine,

I have ~~blatantly copied code from scripts\build\installer\macros.nsh~~ coded functionality similar to the code we had in NSIS for updating.

It seems to install and uninstall correctly, keeping track of everything we told it too.
Even if the user does nasty stuff like install to a game and then uninstall that game or even rename the folder(Which would keep the registry paths pointing to the game..)

On an unrelated <sup>\s</sup> note, guess who has to redownload Fallout 4 and Skyrim?

(Though, as i type this, i realize there may be an edge case in the code. If the user installs bash into Oblivion, and then uninstalls Oblivion, Bash might fail uninstalling because the code to find the game paths checks the registry and i dont believe innosetup saves the value it found at install time. It might, though. I'll test this.)
(Quick testing consisting of renaming game dirs where bash was installed to something else revealed Inno setup: Does nothing. Doesnt complain at all. Goes on it's merry little way. So no edge case, woo!)(This is true for both if the game was uninstalled but the registry keys for it exist, and if they were removed(Tested by *sob* uninstalling my skyrim rip))


Also, there seems to be a weird edge case where bash cant find the game in it's parent directory, but only if Mopy isnt named Mopy.(Discovered since, for testing purposes, i have the installer setup to install to Mopy_ instead of Mopy so i dont bork my own Wrye Bash installs)

The current installer wont overwrite the old installers entry in Add/remove programs, unknown if i'll be able to, if that matters.

I should note that all testing has been done from that branch, which means before any modifications i may have made to the boot process(Which is in a separate branch. Yay!)

Each commit should work, but i cant say for sure. Most commits were tested before being pushed.

I havent yet added most command line options to build.py.

Nothings documented yet, but it will be EXTREMELY easy in the future to add new games to the installer. It currently is, except for the lack of docs. will doc later.

Currently it's setup to be able to install into Skyrim, Skyrim Special Edition, Nehrim, Oblivion, and Fallout 4. So all the games we currently support.(Do we/are we going to support Enderal, come to think of it?)(Then again you're not really supposed to mod that)

I havent added shortcuts yet, but it'll be easy to if thats a blocker.

Extra Locations, Translations, Python Mode, and Hackishly Downloading Prerequisites For Python Mode The User For Some Reason Cant Figure Out How To Download Themselves™ are all unsupported by the current installer.

If it really matters, i'll figure out how to download and install things(I've actually already looked into a bit, InnoSetup has plugins to download files and built in functions for executing files, so), and the same goes for Extra Locations and Python Mode.(For this one, though, it'd be easier on me to make it all or nothing. Either all your installs are Python Mode, or None are, but not a mix.)

So my testing has had everything work fine, i'd like your thoughts though.
If you want to build an installer, the steps are

- Cmd into `Wrye Bash\Building` or otherwise set that to be the working directory.
- Execute build.py
- Wait.
- Everything is now in `Wrye Bash\Building\Release`

All in all, the code should be more reliable installing and removing files, and much easier to add new games too. An arbitrary number of new games, even. Since the components selection screen has SCROLL BARS!!
Something which, i imagine, our current NSIS installer is sorely lacking and would require a mess of code/breaking things to add a new page for yet another new game (:

**Comment by Utumno** [comment]
Please post an issue for the installer, and carry the post above will reply to you there. But for starters yes we need all the functionality of the old installer - removing features is a nono for user experience. Especially the pyton mode and prereqs. An by the way maybe it's the time to have two options for install in preparation of what's to come - one mode like "install Bash in one loactaion" and then options for the different games.

Do you know any of nsis ? If yes I would really appreciate you could fix the bugs reported in the thread, I need the beta out ASAP and a broken installer is a nono (and no time to properly finalize/test this for the beta, but hopefully for 307 proper).

**Comment by iaz3** [comment]
Post an issue here or on my fork or?

"Carry the above post" ?

Sorry, dont know NSIS.

**Comment by iaz3** [comment]
?

**Comment by Utumno** [comment]
Do it on your fork - just carry the post there so we can go on the
discussion there as this issue is for settings refactoring

**Comment by lojack5** [comment]
Since the Window Store games cause a lot of issues with this, I think it might be a good idea to move this up to post-py3 (310?).  We sorely need a better settings system, so we can easily handle issues like this in the future.  I've marked this as 310, but if anyone feels that's not a good idea go ahead and change this.
